client/optimizing.go

```diff
@@ -0,0 +1,309 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/log"
+)
+
+const (
+	defaultRequestTimeout     = time.Second * 5
+	defaultSpeedTestInterval  = time.Minute * 5
+	defaultRequestConcurrency = 2
+)
+
+// NewOptimizingClient creates a drand client that measures the speed of clients
+// and uses the fastest one.
+func NewOptimizingClient(
+	clients []Client,
+	chainInfo *chain.Info,
+	requestTimeout time.Duration,
+	requestConcurrency int,
+	speedTestInterval time.Duration,
+) (Client, error) {
+	if len(clients) == 0 {
+		return nil, errors.New("missing clients")
+	}
+	if chainInfo == nil {
+		return nil, errors.New("missing chain info")
+	}
+	stats := make([]*requestStat, len(clients))
+	now := time.Now()
+	for i, c := range clients {
+		stats[i] = &requestStat{client: c, rtt: 0, startTime: now}
+	}
+	done := make(chan struct{})
+	if requestTimeout <= 0 {
+		requestTimeout = defaultRequestTimeout
+	}
+	if requestConcurrency <= 0 {
+		requestConcurrency = defaultRequestConcurrency
+	}
+	if speedTestInterval <= 0 {
+		speedTestInterval = defaultSpeedTestInterval
+	}
+	oc := &optimizingClient{
+		clients:            clients,
+		stats:              stats,
+		requestTimeout:     requestTimeout,
+		requestConcurrency: requestConcurrency,
+		speedTestInterval:  speedTestInterval,
+		chainInfo:          chainInfo,
+		log:                log.DefaultLogger,
+		done:               done,
+	}
+	go oc.testSpeed()
+	return oc, nil
+}
+
+type optimizingClient struct {
+	sync.RWMutex
+	clients            []Client
+	stats              []*requestStat
+	requestTimeout     time.Duration
+	requestConcurrency int
+	speedTestInterval  time.Duration
+	chainInfo          *chain.Info
+	log                log.Logger
+	done               chan struct{}
+}
+
+type requestStat struct {
+	// client is the client used to make the request.
+	client Client
+	// rtt is the time it took to make the request.
+	rtt time.Duration
+	// startTime is the time at which the request was started.
+	startTime time.Time
+}
+
+type requestResult struct {
+	// result is the return value from the call to Get.
+	result Result
+	// err is the error that occurred from a call to Get (not including context error).
+	err error
+	// stat is stats from the call to Get.
+	stat *requestStat
+}
+
+func (oc *optimizingClient) testSpeed() {
+	for {
```

Comment:
The implementation strategy for testSpeed looks good as it is. No need for change.

FYI, it has one small downside: It queries the clients in bursts (due to the call to parallelGet).
A way to avoid this is to use something like:

i := 0
for {
   time.Sleep(speedTestInterval / len(clients))
   ... test the (i % len(clients))-th client ...
   i++
}